ddd/ddd-02-conn-nonblocking.DISCUSSION.md

```diff
@@ -0,0 +1,167 @@
+
+    CHANGE: The function `BIO_get_fd` is changed to `BIO_get_poll_fd` as the
+        semantics differ; a QUIC implementation will not wish to guarantee any
+        particular properties about the FD returned for polling. In practice we
+        can have `BIO_get_fd` work against QUIC BIOs if we wish, but be
+        considered deprecated.
+
+  - TX can depend on RX and vice versa (SSL_ERROR_WANT_READ/WRITE).
+
+    Q. What if the QUIC stack wants to listen for POLLIN even during
+       a poll caused by WANT_WRITE (more likely) / wants to listen for POLLOUT even
+       during a poll caused by WANT_READ (less likely but possible, e.g. frames
+       transmitted on a timer)?
+
+    PROPOSED SOLUTION: POLLIN should always be listened for.
+
+    CHANGE: Applications should always listen for POLLIN.
+
+    Q. What if the QUIC stack starts to need to read during WANT_WRITE polling?
+    A. Possibly the application should simply always poll for POLLIN.
+       We probably want the QUIC state machine to always handle incoming frames
+       immediately even if TX queues are full so this would make sense.
+
+    PROPOSED SOLUTION: POLLIN should always be listened for.
+
+    CHANGE: If applications do not already test POLLIN during WANT_WRITE they
+        will need to be changed to do so. Arguably they should already do this to
+        handle TLS close notifications etc.
+
+    Q. What if the QUIC stack starts to need to write during WANT_READ polling?
+    A. The QUIC stack should be able to predict at what point in the future
+       it will want to do a spontaneous write so this can be handled by whatever
+       solution is conceived for 'timer issues' below.
+
+    PROPOSED SOLUTION: This is handled by whatever solution we choose for our timer
+        issue. After timer events are handled SSL_read/SSL_write should be
+        tried again which results in a new SSL_ERROR_WANT_READ/WRITE return
+        code. If libssl wants to be notified for both POLLIN and POLLOUT
+        it can return SSL_ERROR_WANT_WRITE due to the decision that POLLIN
+        should always be handled above.
+
+        Applications will need to ensure they retry SSL_read/SSL_write
+        after a timer event (i.e. timer events effectively imply POLLIN|POLLOUT
+        from an application perspective.)
+
+        (Should we also offer the abillity to get a SSL_ERROR_WANT_READ/WRITE
+         as a result of the BIO_pump call?)
+
+  - Timer issues:
+
+    - The QUIC stack may need to randomly send frames (e.g. keepalives)
+      even when the application is not calling SSL_read/SSL_write for a long
+      time (e.g. an hour).
+
+    - So arrangements must be made for some call to libssl to be made
+      periodically.
+
+      PROPOSED SOLUTION: A call, BIO_pump (/SSL_pump) is introduced which can be used
+      when an application wishes to allow libssl to process events without
+      calling read/write functions.
```

Comment:
In DTLS we have `DTLSv1_handle_timeout()` which does this. In openssl/openssl#17184 I introduced `SSL_tick()` to do this.

Reply:
Good point. I'll change this to `BIO_tick`/`SSL_tick`. I think our ideas are the same.